### PR TITLE
Retrieve block root at slot from protoarray to avoid regenerating states

### PR DESCRIPTION
**Pull request by ajsutton**
## PR Description
Nodes on witti were being swamped by beacon block by range requests because Teku was looking up block roots by slot from the historical state roots.  This required walking backwards through states to get to the first request block which is a pathological case during long periods of non-finalisation.

Instead, we can find the block root for a slot from the fork choice protoarray, for anything past finalisation (which is all `RecentChainData` is intended to provide).

Limit the number of blocks included in BeaconBlocksByRange responses.

## Fixed Issue(s)
Expecting this to help with if not eliminate #2149 - we have a ton of close futures because we're still generating results for the requests they made.

## Documentation

- [x] I thought about documentation and added the `documentation` label to this PR if updates are required.

### CODE REVIEW

File: networking/eth2/src/test/java/tech/pegasys/teku/networking/eth2/rpc/beaconchain/methods/BeaconBlocksByRangeMessageHandlerTest.java [hunk 1]
```diff
@@ -269,6 +272,23 @@ public void shouldRejectRequestWhenStepIsZero() {
     verifyNoMoreInteractions(combinedChainDataClient);
   }
 
+  @Test
+  void shouldLimitNumberOfBlocksReturned() {
+    final RequestState requestState =
+        new RequestState(
+            new BeaconBlocksByRangeRequestMessage(
+                UnsignedLong.ZERO, UnsignedLong.valueOf(500), UnsignedLong.ONE),
+            UnsignedLong.valueOf(10000),
```
Comment (cemozerr) [inline networking/eth2/src/test/java/tech/pegasys/teku/networking/eth2/rpc/beaconchain/methods/BeaconBlocksByRangeMessageHandlerTest.java:283]: nit: We could make the test more explicit by saying count = Constants.MAX_BLOCK_BY_RANGE_REQUEST_SIZE + 1

File: storage/src/main/java/tech/pegasys/teku/storage/client/RecentChainData.java [hunk 2]
```diff
@@ -304,87 +308,17 @@ public boolean containsBlock(final Bytes32 root) {
...
-
-      // Pull an older block to search
-      final UnsignedLong oldestQueryableSlot = currentBlock.getSlot().minus(slotsPerHistoricalRoot);
-      final UnsignedLong olderSlot = max(youngestQueryableSlot, oldestQueryableSlot);
-      final Bytes32 olderRoot = get_block_root_at_slot(currentBlock.getState(), olderSlot);
-      currentBlock = store.getBlockAndState(olderRoot).orElse(null);
-    }
-
-    return Optional.empty();
+  public Optional<Bytes32> getBlockRootBySlot(
```
Comment (cemozerr) [inline storage/src/main/java/tech/pegasys/teku/storage/client/RecentChainData.java:319]: Are we okay with not being able to get any block root before finalization?

Comment (cemozerr) [reply storage/src/main/java/tech/pegasys/teku/storage/client/RecentChainData.java:319]: If we do want to use ProtoArray for collecting block roots pre-finalization, we could modify it in such a way that `findHead` (and thus `computeDeltas`) would not traverse through ProtoNodes pre-finalization, and stop at the finalization slot, but we'd still have access to information on nodes pre-finalization. This way ForkChoice could stay efficient.

**Review by cemozerr — APPROVED** [review]
LGTM.

File: storage/src/main/java/tech/pegasys/teku/storage/client/RecentChainData.java (hunk 2 above)
Comment (ajsutton) [reply storage/src/main/java/tech/pegasys/teku/storage/client/RecentChainData.java:319]: There was always a limit to how far back we could go with `RecentChainData` anyway so I think it's fine if it cuts off at the finalised block.  We don't have the blocks or states before that anyway.

File: networking/eth2/src/test/java/tech/pegasys/teku/networking/eth2/rpc/beaconchain/methods/BeaconBlocksByRangeMessageHandlerTest.java (hunk 1 above)
Comment (ajsutton) [reply networking/eth2/src/test/java/tech/pegasys/teku/networking/eth2/rpc/beaconchain/methods/BeaconBlocksByRangeMessageHandlerTest.java:283]: Done.

**Pull request merged**